### PR TITLE
feat: sanity checks

### PR DESCRIPTION
**Pull request by Dev-Doggo**
Closes: truflation/truf-claim-page#224

### CODE REVIEW

**Comment by Dev-Doggo** [comment]
@markholdex I really don't like how this code looks imo, but if it's something we want to implement, sure 🤷‍♂️ 

cc: @zolotokrylin

**Review by lewis999bits — DISMISSED** [review]

File: src/staking/StakingRewards.sol [hunk 1]
```diff
@@ -45,6 +45,8 @@ contract StakingRewards is IStakingRewards, ReentrancyGuard, Pausable, Ownable2S
     /* ========== CONSTRUCTOR ========== */
 
     constructor(address _rewardsDistribution, address _rewardsToken, address _stakingToken) {
+        require(IERC20(_rewardsToken).totalSupply() > 0, "Invalid token");
```
Comment (lewis999bits) [inline src/staking/StakingRewards.sol:48]: @Dev-Doggo Please add sanity check for zero addresses as you did in `VirtualStakingRewards`

**Review by lewis999bits — CHANGES_REQUESTED** [review]

File: src/TrufPartner.sol [hunk 2]
```diff
@@ -80,6 +83,14 @@ contract TrufPartner is Ownable2Step {
             revert ZeroAddress();
         }
 
+        if (IERC20(_trufToken).totalSupply() != TRUF_TOTAL_SUPPLY) {
+            revert InvalidToken();
+        }
```
Comment (zolotokrylin) [inline src/TrufPartner.sol:88]: You don't need this. A vesting contract is not supposed to check if the supply of the token is the correct supply. The token you pass in the constructor already has the correct supply.

What the auditor meant, as I understand, is that you check the token supply when you call a function, where you supply the token amount to be used. So when you have for example ABC token with a total supply of 100 tokens, you must not allow the vesting admin to set 1,000 tokens for vesting.

Or am I missing something?

Comment (Dev-Doggo) [reply src/TrufPartner.sol:88]: @zolotokrylin If you check the references the auditor left they all link to the constructor arguments, this is what I understood from it even though as I said previously I've never seen this implemented in any other protocol.
I still think this is completely unnecessary to implement

Also I think implementing what you suggested for the vesting contract is going to be quite costly in terms of user gas, I'd need to look into it further if we actually want to implement that, please let me know!

**Comment by lewis999bits** [comment]
@Dev-Doggo I just reviewed the code and for me I cannot approve or reject this PR.
I've never seen this kind of check before and those lines cannot cover the sanity check 100%.
Is this what audit team really want to do? 
cc @markholdex @zolotokrylin

**Comment by Dev-Doggo** [comment]
> @Dev-Doggo I just reviewed the code and for me I cannot approve or reject this PR. I've never seen this kind of check before and those lines cannot cover the sanity check 100%. Is this what audit team really want to do?

@lewis999bits As far as I understand it yes, the point is that if the address passed in is incorrect the `.totalSupply`/`.balanceOf` calls will fail as it will not return any data/revert.

I don't think it's necessary to implement this though, it was quick to do so which is why I implemented it and we can discuss if it's something we want to merge.

**Review by lewis999bits — APPROVED** [review]

**Comment by lewis999bits** [comment]
@Dev-Doggo I don't know if I'm out of this context, but the sanity check the audit team supposed is more likely needed for the dynamic contract deployment (f.e factory contract). Our smart contracts are one-click deployment and the checks can be done whether manually or via script. This PR just increase unnecessary codes and increase deployment gas fee which is painful for us. And in terms of architecture, it makes our code poor.

**Comment by Dev-Doggo** [comment]
> @Dev-Doggo I don't know if I'm out of this context, but the sanity check the audit team supposed is more likely needed for the dynamic contract deployment (f.e factory contract). Our smart contracts are one-click deployment and the checks can be done whether manually or via script. This PR just increase unnecessary codes and increase deployment gas fee which is painful for us. And in terms of architecture, it makes our code poor.

I agree with you, some of those points I also raised with the auditor in the initial Problem as well, they said they can remove this from the final report even with no fixes.

cc @zolotokrylin

**Comment by markholdex** [comment]
@Dev-Doggo @lewis999bits alright. Let's close without merging.

**Comment by markholdex** [comment]
Don't delete the branch yet

**Pull request closed without merging**